obj/generators.h

```diff
@@ -4,75 +4,53 @@
 #include <initializer_list>
 #include <iostream>
 #include <random>
 
 namespace generators {
-struct RandomGen {
-  virtual ~RandomGen(){};
-  virtual float getNext() = 0;
-};
 
-struct UniformRandomGen : public RandomGen {
-  UniformRandomGen(float min, float max)
-      : engine_(std::random_device()()), dist_(min, max){};
-  float getNext() override { return dist_(engine_); }
+struct RandomRayOffseter {
+public:
+  core::Ray offsetRay(const core::Ray &ray) {
+    return core::Ray(ray.origin(),
+                     ray.direction() + core::Vec3(getNextAxisOffset(),
+                                                  getNextAxisOffset(), 0));
+  }
 
-  std::mt19937_64 engine_;
-  std::uniform_real_distribution<float> dist_;
+protected:
+  virtual float getNextAxisOffset() const = 0;
 };
 
-struct FakeRandomGen : public RandomGen {
-  float getNext() override { return 0; }
+class RayFactory {
+public:
+  // generates ray with properties determinated by each class
+  // example: Speaker source is a membrane that is generating
+  // rays that have origins where membrane is situated and direction
+  // parpendicular to surface where membrane is situated.
+  virtual bool genRay(core::Ray *ray) = 0;
 };
 
-// This object is generating rays with the origin
-// of the point source position and direction calculated
-// from leftCorner position + x * Vec3(1, 0, 0) + y * Vec3(0, 1, 0)
-// and origin of the source position
-class PointSource final {
+// generates rays with origin at PointSource origin
+// and direction along Z axes down, with offset on XY
+// determinated by sampleSize until numOfRay number is reached
+class PointSpeakerRayFactory : public RayFactory {
 public:
-  PointSource() = delete;
-  explicit PointSource(
-      float freq, int numOfRaysPerRow, float SampleSize, RandomGen *randomGen,
-      core::Vec3 origin = core::Vec3(0, 0, constants::kDefaultSimulationRadius))
-      : frequency_(freq), numOfRaysPerRow_(numOfRaysPerRow),
-        sampleSize_(SampleSize), origin_(origin), randomGen_(randomGen) {
-    updateSampleSize();
-  }
-
-  bool operator==(const PointSource &other) const;
-  friend std::ostream &operator<<(std::ostream &os,
-                                  const PointSource &pointSource);
-
-  void updateSampleSize();
-  core::Ray generateRay(int xIter, int yIter);
-
-  float frequency() const;
-  void setFrequency(float freq);
-
-  float sampleSize() const;
-  void setDiffusorSize(float diffusorSize);
-
-  size_t numOfRaysPerRow() const;
-  void setNumOfRaysPerRow(int raynum);
-
-  core::Vec3 origin() const;
-  void setOrigin(const core::Vec3 &point);
+  PointSpeakerRayFactory(int numOfRays, Model *model);
 
-  core::Vec3 getLeftCorner() const;
-  void setLeftCorner(const core::Vec3 &point);
+  [[nodiscard]] bool genRay(core::Ray *ray) override;
 
 private:
-  core::Vec3 origin_, leftCorner_;
-  float frequency_, sampleSize_;
-  size_t numOfRaysPerRow_;
+  int numOfRays_, currentRayNum_;
+  Model *model_;
```

Comment:
Do you need to store this here? I thought you only need to keep it in c-tor to get some values out.

Reply:
true, fixed